deepchecks/nlp/checks/data_integrity/frequent_substrings.py

```diff
@@ -0,0 +1,227 @@
+        minimum threshold for tagging a substring as frequent
+    min_ngram: int, default: 4
+        minimum n for the n-gram extraction
+    epsilon: float, default: 0.02
+        maximum difference for filtering long substrings
+    significant_threshold: float, default: 0.3
+        all the samples above the frequency would be returned
+
+    """
+
+    def __init__(
+        self,
+        n_to_show: int = 5,
+        n_samples: int = 10_000_000,
+        random_state: int = 42,
+        min_threshold: float = 0.05,
+        min_ngram: int = 4,
+        epsilon=0.02,
+        significant_threshold: float = 0.3,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.n_to_show = n_to_show
+        self.n_samples = n_samples
+        self.random_state = random_state
+        self.min_threshold = min_threshold
+        self.min_ngram = min_ngram
+        self.epsilon = epsilon
+        self.significant_threshold = significant_threshold
+
+    @staticmethod
+    def _extract_ngrams(text, n):
+        words = text.split()
+        return [' '.join(words[i:i + n]) for i in range(len(words) - n + 1)]
+
+    @staticmethod
+    def _get_peak_cut_ind(df, min_diff=0.05):
+        df['diff'] = df['Frequency'].diff()
+        df[df['diff'] <= min_diff]['diff'] = 0
+        max_peak = df['diff'].min()
+        if max_peak == 0:
+            return df
+        return list(df[df['diff'] == max_peak].index)[0]
+
+    def _get_ngram_frequencies_and_indexes(self, data, n, num_samples):
+        ngram_info = defaultdict(lambda: {'freq': 0, 'original_indexes': set(), 'new_indexes': set()})
+        qualifying_samples = set()
+
+        ngrams = set()
+        for item in data:
+            for ngram in self._extract_ngrams(item[1], n):
+                ngrams.add(ngram)
+
+        for ngram in ngrams:
+            for index, item in enumerate(data):
+                if ngram in item[1]:
+                    ngram_info[ngram]['freq'] += 1
+                    ngram_info[ngram]['original_indexes'].add(item[0])
+                    ngram_info[ngram]['new_indexes'].add(index)
+
+        for ngram in ngram_info:
+            ngram_info[ngram]['original_indexes'] = list(ngram_info[ngram]['original_indexes'])
+            ngram_info[ngram]['new_indexes'] = list(ngram_info[ngram]['new_indexes'])
+            ngram_info[ngram]['freq'] /= num_samples
+            if ngram_info[ngram]['freq'] >= self.min_threshold:
+                for index in ngram_info[ngram]['new_indexes']:
+                    qualifying_samples.add(index)
+
+        return ngram_info, qualifying_samples
+
+    def _get_frequent_ngrams(self, data):
+        n = self.min_ngram
+        final_results = {}
+        num_samples = len(data)
+        while data:
+            result, qualifying_samples = self._get_ngram_frequencies_and_indexes(data, n, num_samples)
+            result = {ngram: info for ngram, info in result.items() if info['freq'] >= self.min_threshold}
```

Comment:
why filter here for ` >= self.min_threshold`? Couldn't we do that inside the for ngram in ngram_info: loop?